hknweb/candidate/views.py

```diff
@@ -4,18 +4,34 @@
 from django.core.mail import EmailMultiAlternatives
 from django.template.loader import render_to_string
 from django.contrib import messages
-from django.contrib.auth.decorators import login_required
+from django.contrib.auth.decorators import login_required, user_passes_test
 from django.utils.decorators import method_decorator
 from django.conf import settings
 from django.contrib.staticfiles.finders import find
 from django.contrib.staticfiles.templatetags.staticfiles import static
+from django.conf import settings
 from random import randint
 
 from .models import OffChallenge, Announcement
 from .forms import ChallengeRequestForm, ChallengeConfirmationForm
 
+# decorators
+
+# used for things only officers and candidates can access
+def check_account_access(func):
+    def check_then_call(request, *args, **kwargs):
+        if not is_cand_or_officer(request.user):
+            return render(request, "errors/401.html", status=401)
+        return func(request, *args, **kwargs)
+    return check_then_call
+
+
+# views
+
 # Candidate portal home
 @method_decorator(login_required(login_url='/accounts/login/'), name='dispatch')
+@method_decorator(check_account_access, name='dispatch')
```

Comment:
The dispatch function is inherited from the parent class. I used method decorators because that's what the example I found did, and it also that way you can use the same decorator for methods and functions.

Reply:
Could you post a link to the example?

Reply:
It looks like there are less-verbose alternatives to these decorators, namely mixin classes to inherit from: https://docs.djangoproject.com/en/2.2/topics/auth/default/#django.contrib.auth.mixins.LoginRequiredMixin

I can rewrite this code with mixin inheritance in a separate pull request.